internal/controllers/topology/cluster/cluster_controller.go

```diff
@@ -202,9 +204,32 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 func (r *Reconciler) reconcile(ctx context.Context, s *scope.Scope) (ctrl.Result, error) {
 	var err error
 
+	// Get ClusterClass.
+	clusterClass := &clusterv1.ClusterClass{}
+	key := client.ObjectKey{Name: s.Current.Cluster.Spec.Topology.Class, Namespace: s.Current.Cluster.Namespace}
+	if err := r.Client.Get(ctx, key, clusterClass); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve ClusterClass %s", s.Current.Cluster.Spec.Topology.Class)
+	}
+
+	s.Blueprint.ClusterClass = clusterClass
+	// If the ClusterClass `metadata.Generation` doesn't match the `status.ObservedGeneration` requeue as the ClusterClass
+	// is not up to date.
+	if clusterClass.GetGeneration() != clusterClass.Status.ObservedGeneration {
+		return ctrl.Result{}, errors.Errorf("ClusterClass %s not up to date. Requeuing", clusterClass.Name)
```

Comment:
Now that I'm looking at the new code. Another idea on top. We could also do the following:
* `return ctrl.Result{}, nil` here
  * we don't have to requeue as an event is triggered when `clusterClass.Status.ObservedGeneration` is written on the ClusterClass
* handle the generation case in `reconcileConditions` as well
  * Basically add a check after we check `reconcileErr` and write the TopologyReconciledCondition with whatever error we want there

Downside:
* we have to check for generation in two places
Upside:
* we don't produce additional unnecessary requeues (including exponential backoff)
* we don't produce "error reconciles" in metrics
* we still get the right information into the condition for users

Reply:
Looks good to me